iterative/aws/provider.go

```diff
@@ -188,26 +192,48 @@ func ResourceMachineCreate(ctx context.Context, d *schema.ResourceData, m interf
 	sgID = *sgDesc.SecurityGroups[0].GroupId
 	vpcID = *sgDesc.SecurityGroups[0].VpcId
 
-	subDesc, err := svc.DescribeSubnets(ctx, &ec2.DescribeSubnetsInput{
+	// https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ec2#Client.DescribeSubnets
+	// default Subnet selection
+	subnetOptions := &ec2.DescribeSubnetsInput{
 		Filters: []types.Filter{
 			{
 				Name:   aws.String("vpc-id"),
 				Values: []string{vpcID},
 			},
 		},
-	})
+	}
+	// use availability zone from user
+	if availabilityZone != "" {
+		subnetOptions.Filters = append(subnetOptions.Filters, types.Filter{
+			Name:   aws.String("availability-zone"),
+			Values: []string{availabilityZone},
+		})
+	}
+	// use exact subnet-id from user
+	if subnetId != "" {
+		subnetOptions.Filters = append(subnetOptions.Filters, types.Filter{
+			Name:   aws.String("subnet-id"),
+			Values: []string{subnetId},
+		})
+	}
+	subDesc, err := svc.DescribeSubnets(ctx, subnetOptions)
 	if err != nil {
 		return decodeAWSError(region, err)
 	}
 	if len(subDesc.Subnets) == 0 {
-		return errors.New("no subnets found")
+		return errors.New("No Subnet found")
 	}
 	var subnetID string
-	for _, subnet := range subDesc.Subnets {
-		if *subnet.AvailableIpAddressCount > 0 && *subnet.MapPublicIpOnLaunch {
-			subnetID = *subnet.SubnetId
-			break
+	// bypass with user provided ID
+	if subnetId == "" {
```

Comment:
I think that this code is not correct. Leave it as it was.
The reason is that we need the subnet to be have a publicIp. The check is still in place and a must. 
The code is returning the first subnet that have a public IP but because you are adding a filter in your Describe priorly your subnets list will be composed by one element at most.
To that element we will do the publicip check

Reply:
I want to bypass the public IP check by using a specific subnet-id.

My "hypothetical" reason is some Org's may require certain complex requirements on public-facing instances even with secure defaults with key-based authentication via ssh which is the only port exposed.

By providing a specific subnet-id the user should verify the instance is routable:

```
[public]        [10.0.0.0/8]
[Github]        [Self-Hosted ghar]
#  <------------ waits for jobs
push event ----> gets job                   [10.0.0./8]
                 cml runner --------------> [ CML Instance]
                                            [ephermal github runner]
# <---------------------------------------- [Training]
                                            [Instance tear down]
#  <------------ waits for jobs                

```

If want to try and have a quick meeting I can try and better explain the use case for some of these features I have been adding.

Reply:
> I want to bypass the public IP check by using a specific subnet-id.

That would work for the `machine` but not for the `runner`
We need to check the runner readiness via SSH.

Reply:
But If `cml runner` is run from within the same subnet is will work 😸 which is the intention here. I would argue that if a user supplies an exact subnet/vpc to use the onus is on them to ensure the networking allows the connection from where they are invoking `cml runner` and thus the readiness check via SSH will work.

It should totally be documented that you could shoot yourself and create an unroutable instance.

I'm happy to connect so I can explain the use case further or alternatively, I can check our NDA status and can send a teams meeting to @casperdcl to show you guys what I mean for this environment/use-case to accomplish.

Reply:
Also @DavidGOrtega @0x2b3bfa0 my choice of `subnetId` may have been poor with `subnetID` also existing. Where `subnetId` is my user provided option, an example being: `"subnet-06ee5a8e940002c2d"` while `subnetID` is the original value derived from the `svc.DescribeSubnets` call

Reply:
Ahhhhh! got it. As you say this would work for on-premise instances! 🙏 

> my choice of subnetId may have been poor with subnetID also existing.

Originally I was suggesting remove the else assigning subnetID with subnetId at least. Then I realised about the publicip need.

Reply:
Exactly, I guess I was avoiding the terminology "on-premise" since this is still happening on AWS but it is "on-premise" in the sense that it involves a hybrid-multi-cloud buzzword setup.